src/main/kotlin/uk/gov/justice/digital/hmpps/prisonerfromnomismigration/incidents/IncidentsReconciliationService.kt

```diff
@@ -72,6 +96,89 @@ class IncidentsReconciliationService(
       ),
     )
   }.getOrNull()
+
+  suspend fun checkOpenIncidentsMatch(agencyId: String, openIncidentCount: Long): List<MismatchOpenIncident> {
+    return openIncidentCount.asPages(pageSize).flatMap { page ->
+      val openIncidentIds = getOpenIncidentsForPage(agencyId, page)
+      openIncidentIds.mapNotNull { checkOpenIncidentMatch(it.incidentId) }
+    }
+  }
+
+  private suspend fun getOpenIncidentsForPage(agencyId: String, page: Pair<Long, Long>) =
+
+    runCatching { nomisIncidentsApiService.getOpenIncidentIds(agencyId, page.first, page.second).content }
+      .onFailure {
+        telemetryClient.trackEvent(
+          "incidents-reports-reconciliation-mismatch-page-error",
+          mapOf(
+            "page" to page.first.toString(),
+          ),
+        )
+        log.error("Unable to match entire page of incidents: $page", it)
+      }
+      .getOrElse { emptyList() }
+      .also { log.info("Page requested: $page, with ${it.size} open incidents") }
+
+  private suspend fun checkOpenIncidentMatch(nomisOpenIncidentId: Long): MismatchOpenIncident? = runCatching {
+    val (nomisOpenIncident, dpsOpenIncident) =
+      withContext(Dispatchers.Unconfined) {
+        async { nomisIncidentsApiService.getIncident(nomisOpenIncidentId) } to
+          async { incidentsService.getIncidentDetailsByNomisId(nomisOpenIncidentId) }
+      }.awaitBoth()
```

Comment:
Id suggest you add the doApiCallWithRetries{} wrapper (e.g. see the locations reconciliation) as there can be a lot of network failures.

Reply:
Good idea - I did that in a few other places.